src/memkind_memtier.c

```diff
@@ -1189,14 +1192,63 @@ MEMKIND_EXPORT void *memtier_malloc(struct memtier_memory *memory, size_t size)
     void *ptr;
     uint64_t data;
 
-    ptr = memtier_kind_malloc(memory->get_kind(memory, size, &data), size);
+    dont_mmap = true;
+    memkind_t kind = memory->get_kind(memory, size, &data);
+    ptr = memtier_kind_malloc(kind, size);
     memory->post_alloc(data, ptr, size);
     memory->update_cfg(memory);
     print_memory_statistics(memory);
 
+    dont_mmap = false;
     return ptr;
 }
 
+void *sys_mmap(void *addr, size_t length, int prot, int flags, int fd, off_t off)
+{
+    long ret = syscall(SYS_mmap, addr, length, prot, flags, fd, off);
+    if (ret == -EPERM && !off && (flags&MAP_ANON) && !(flags&MAP_FIXED))
+        ret = -ENOMEM;
+    if (ret > -4096 && ret < 0) {
+        errno = -ret;
+        return MAP_FAILED;
+    }
+
+    return (void*)ret;
+}
+
+int munmap(void *addr, size_t length)
+{
+    long ret = syscall(SYS_munmap, addr, length);
+    if (!ret)
+        return 0;
+    errno = -ret;
+    return -1;
+}
+
+#include <pthread.h>
+pthread_mutex_t mutex;
+
+MEMKIND_EXPORT void* memtier_mmap(struct memtier_memory *memory, void *addr, size_t length, int prot, int flags, int fd, off_t offset)
+{
+    if ((memory == NULL) || dont_mmap)
+        return sys_mmap(addr, length, prot, flags, fd, offset);
+
+    // prevent recursive mmap
+    dont_mmap = true;
+    
+    pthread_mutex_lock(&mutex);
+    void* ret = memtier_malloc(memory, length);
```

Comment:
Isn't malloc supposed to be thread safe?

We only use variables specified as arguments here, do we use the mutex to protect them?

Reply:
This does not support splittings big mmaps into multiple smaller remaps on different regions... I thought this was supposed to be the key feature